Add a level of indirection to config file inclusion

bensze01 opened the pull request:
Split the config files into one that's purely made of `#defines`, intended to be edited by end users, and one that has all the necessary preprocessor logic, extracted from the former config file, and other source files.

Closes #3960

## Status
**Ready**

## Requires Backporting
No

## Migrations
The config file was renamed split and renamed, and a mandatory version symbol was added.

## Todos
- [x] Documentation
- [x] Changelog updated

mstarzyk-mobica commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -0,0 +1,67 @@
+  *  limitations under the License.
+  */
+
+#ifndef MBEDTLS_MBEDTLS_CONFIG_H
```
duplicated `MBEDTLS_`

mstarzyk-mobica commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -36,13 +36,26 @@
  * Allow user to override any previous default.
  *
  */
+#if defined(MBEDTLS_USER_CONFIG_VERSION)
```
Above comment is addressing the code in lines 53-54

bensze01 commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -0,0 +1,67 @@
+  *  limitations under the License.
+  */
+
+#ifndef MBEDTLS_MBEDTLS_CONFIG_H
```
Yes, I've duplicated `MBEDTLS_`, since our convention for header include guards has been, from what I can gather `NAME_OF_FOLDER_NAME_OF_HEADER`. So `mbedtls/mbedtls_c.onfig.h` translates to an include guard of `MBEDTLS_MBEDTLS_CONFIG_H`

bensze01 commented on scripts/generate_query_config.pl:
```diff
@@ -64,9 +64,6 @@
         my $name = $2;
 
-        # Skip over the macro that prevents multiple inclusion
-        next if "MBEDTLS_CONFIG_H" eq $name;
```
I've removed the include guard from `config.h` / `mbedtls_config_defines.h`, (since the guard against multiple inclusion is now in `mbedtls_config.h`.

gabor-mezei-arm reviewed (changes requested):
The 3.0 migration guide should also be updated how the config file inclusions are changed.

gabor-mezei-arm commented on docs/3.0-migration-guide.d/split_config.md:
```diff
@@ -0,0 +1,16 @@
+`mbedtls_config.h` is intended to be included from C code directly, while
+`mbedtls_config_defines.h` is intended to be edited by end users whishing to
+change the build configuration, and should generally only be included from
+`mbedtls_config_defines.h`. This is because all the preprocessor logic has been
```
Please correct:
`mbedtls_config_defines.h` -> `mbedtls_config.h`

gabor-mezei-arm commented on ChangeLog.d/split-config.txt:
```diff
@@ -0,0 +1,12 @@
+     mbedtls_config.h is intended to be included from C code directly, while
+     mbedtls_config_defines.h is intended to be edited by end users whishing to
+     change the build configuration, and should generally only be included from
+     mbedtls_config_defines.h.
```
Please correct:
`mbedtls_config_defines.h` -> `mbedtls_config.h`

gabor-mezei-arm reviewed (approved):
Looks good to me

bensze01 commented:
I've rebased this PR on top of `development`.

gabor-mezei-arm reviewed (dismissed):
Looks good to me

bensze01 commented:
I've rebased the PR again.

mstarzyk-mobica commented:
I don't have a local copy of the pre-rebase branch so I don't know if the rebase was done correctly.
@bensze01 can you share the pre-rebase version in a fork?

gilles-peskine-arm commented:
@mstarzyk-mobica You can get a local copy with `git fetch origin 3b37ab3190c618eb1594fcc4e2155e48d7148bc3` (replace `origin` by the name of the remote if you chose a different one), or `git fetch origin 3b37ab3190c618eb1594fcc4e2155e48d7148bc3:pr_4589-5` to give it a local branch name. GitHub started supporting this a while ago.

bensze01 commented:
As per discussion with @gilles-peskine-arm, I'll stop rebasing this until the design review is done.

davidhorstmann-arm commented on ChangeLog.d/split-config.txt:
```diff
@@ -0,0 +1,12 @@
+Changes
+   * config.h has been split into mbedtls_config.h and mbedtls_config_defines.h
+     mbedtls_config.h is intended to be included from C code directly, while
+     mbedtls_config_defines.h is intended to be edited by end users whishing to
```
Nit: 'whishing' -> 'wishing'

davidhorstmann-arm commented on docs/3.0-migration-guide.d/split_config.md:
```diff
@@ -0,0 +1,16 @@
+
+`config.h` was split into `mbedtls_config.h` and `mbedtls_config_defines.h`.
+`mbedtls_config.h` is intended to be included from C code directly, while
+`mbedtls_config_defines.h` is intended to be edited by end users whishing to
```
Once again, 'wishing'

davidhorstmann-arm reviewed (changes requested):
Couple of spelling nits, otherwise LGTM

davidhorstmann-arm reviewed (approved):
LGTM - thanks!

gilles-peskine-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -0,0 +1,67 @@
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if !defined(MBEDTLS_CONFIG_VERSION) || \
```
I'm not sold on _requiring_ `MBEDTLS_CONFIG_VERSION` to be set. Why don't we allow config files that don't set it, for users who just want certain features to be enabled and don't care about details? It's not like we're planning to change `MBEDTLS_CONFIG_VERSION` often anyway.

In other words, I think that a config file with `MBEDTLS_CONFIG_VERSION` undefined should mean “I want this configuration, fill in any blanks in a sensible way”, while a config file with `MBEDTLS_CONFIG_VERSION` set should mean “I want this configuration exactly and I'm very sensitive about backward compatibility and not getting additional stuff that could increase code size”.

gilles-peskine-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -0,0 +1,67 @@
+#endif
+
+#if !defined(MBEDTLS_CONFIG_VERSION) || \
+    MBEDTLS_CONFIG_VERSION != 1
```
I think it's confusing to have version 1 be the version introduced in Mbed TLS 3.0. Should we make this version 3? Or version 0x030000? Or version 202106?

ronald-cron-arm replied:
What about for the version of the configuration to just follow the version of the library ? That way we don't have two different versions for the config and the library.
At each release we update in the editable config.h the config version to that of the library version.
Then for a library version  of major version M >=  3, only configuration versions MBEDTLS_CONFIG_VERSION lower than the library version and with the same major version are acceptable. If MBEDTLS_CONFIG_VERSION is strictly lower than the library version MBEDTLS_VERSION_NUMBER the resulting configuration is as if the library version was equal to MBEDTLS_CONFIG_VERSION.

gilles-peskine-arm reviewed (changes requested):
I have done a partial design review and I have a few comments.

ronald-cron-arm commented on ChangeLog.d/split-config.txt:
```diff
@@ -0,0 +1,12 @@
+Changes
+   * config.h has been split into mbedtls_config.h and mbedtls_config_defines.h
+     mbedtls_config.h is intended to be included from C code directly, while
```
+1 from me for mbedtls/base.h. Then I think we should rename the current config.h to mbedtls_config.h (and not mbedtls_config_defines.h): we just add the "mbedtls_" prefix to ease the integration in other projects.

bensze01 replied:
I'm planning to go with `mbedtls/build_info.h` for the includable config, and `mbedtls/mbedtls_config.h` for the editable one.

bensze01 commented:
@gilles-peskine-arm @ronald-cron-arm I think I've addressed all your feedback

I'm also thinking about just ripping out MBEDTLS_USER_CONFIG_VERSION and the logic that checks if it matches MBEDLTS_CONFIG_VERSION altogether.

Any thoughts about that?

bensze01 commented:
I uploaded the [previous reviewed version](https://github.com/bensze01/mbedtls/tree/split_config_1) of this pr, and the version [simply rebased on top of development](https://github.com/bensze01/mbedtls/tree/split_config_rebase_1
), with none of the news changed  to my repo, for the reviewers convenience

gilles-peskine-arm commented:
> I'm also thinking about just ripping out MBEDTLS_USER_CONFIG_VERSION

Well, [I still don't understand what it's for](https://github.com/ARMmbed/mbedtls/pull/4589#discussion_r650200414). Can you please explain when it would come into play?

bensze01 commented:
> > I'm also thinking about just ripping out MBEDTLS_USER_CONFIG_VERSION
> 
> Well, [I still don't understand what it's for](https://github.com/ARMmbed/mbedtls/pull/4589#discussion_r650200414). Can you please explain when it would come into play?

I introduced it because I was unsure of how to handle the interaction of MBEDTLS_CONFIG_VERSION and MBEDTLS_USER_CONFIG.

Since handling two config files with independent versions is a pain, and I originally made a version symbol in the config files mandatory, I decided to make a corresponding mandatory version symbol in the MBEDTLS_USER_CONFIG file, and throw an error if the versions mismatched, hopefully more informative error than interpreting MBEDTLS_USER_CONFIG under whatever version the other config file specified.

This way I could just easily compare them and make sure that they were exactly the same. This also made a bit more sense while config versions were independent of the mbedtls version.

gilles-peskine-arm commented on ChangeLog.d/split-config.txt:
```diff
@@ -0,0 +1,12 @@
+     mbedtls_config.h.
+   * The handling of MBEDTLS_CONFIG_FILE has been moved into mbedtls_config.h.
+   * Mandatory version symbols MBEDTLS_CONFIG_VERSION and
+     MBEDTLS_USER_CONFIG_VERSION were introduced for MBEDTLS_CONFIG_FILE and
```
Looking at what's implemented in f78b0f8225be2d68ec4226535082c57694dca99c:

* `MBEDTLS_USER_CONFIG_VERSION` must be either left undefined, or defined in `MBEDTLS_USER_CONFIG_FILE`.
* `MBEDTLS_USER_CONFIG_VERSION` is never mandatory.
* The sole effect of `MBEDTLS_USER_CONFIG_VERSION` is to cause an error if it's set to a value that's different from `MBEDTLS_CONFIG_VERSION`.
* A similar effect can be achieved by including `#define MBEDTLS_CONFIG_VERSION` in the user configuration file: if it's set to a different token sequence from the previously defined value of `MBEDTLS_CONFIG_VERSION`, most compilers will complain.

So…

> I'm also thinking about just ripping out MBEDTLS_USER_CONFIG_VERSION and the logic that checks if it matches MBEDLTS_CONFIG_VERSION altogether.

Yes, please do, I don't think it's useful.

bensze01 commented:
@gilles-peskine-arm I went ahead and removed it.

gilles-peskine-arm commented on docs/3.0-migration-guide.d/split_config.md:
```diff
@@ -0,0 +1,17 @@
+`build_info.h`. This is because all the preprocessor logic has been moved
+into `build_info.h`, including the handling of the `MBEDTLS_CONFIG_FILE`
+macro.
+
```
I think we can be clearer here:

* In code, use `#include <mbedtls/build_info.h>`. Don't include `mbedtls/config.h` and don't refer to `MBEDTLS_CONFIG_FILE`.
* In build tools, edit `mbedtls_config.h`, or edit `MBEDTLS_CONFIG_FILE` as before.
* If you had a tool that parsed the library version from `include/mbedtls/version.h`, this has moved to `include/mbedtls/build_info.h`. From C code, both headers now define the `MBEDTLS_VERSION_xxx` macros.

Also, if you have a custom configuration file:

* Don't include `check_config.h` anymore.
* Don't define `MBEDTLS_CONFIG_H` anymore.

gilles-peskine-arm commented on include/mbedtls/build_info.h:
```diff
@@ -36,6 +59,12 @@
 #include MBEDTLS_CONFIG_FILE
 #endif
 
+#if defined(MBEDTLS_CONFIG_VERSION) && ( \
+    MBEDTLS_CONFIG_VERSION < 0x03000000 || \
+    MBEDLTS_CONFIG_VERSION > MBEDTLS_VERSION_NUMBER )
```
It's a bit awkward that until the release, people won't be able to try out defining `MBEDLTS_CONFIG_VERSION`, since no version is currently acceptable. But that's only going to last a few days, so I guess it doesn't matter.

gilles-peskine-arm replied:
```suggestion
#if defined(MBEDTLS_CONFIG_VERSION) && ( \
    MBEDTLS_CONFIG_VERSION < 0x03000000 || \
    MBEDTLS_CONFIG_VERSION > MBEDTLS_VERSION_NUMBER )
```

gilles-peskine-arm reviewed (changes requested):
Looks almost good to me.

I haven't played as much with the new arrangement as I would have liked to, but I think the configuration version mechanism is flexible enough that it's likely we'll have good solutions to future problems where the best solution is to change the meaning of a configuration option, which was the primary motivation for this task. So that's a great design! Just a few typos to fix and documentation to clarify.

bensze01 commented:
I think I've addressed all of the feedback from @gilles-peskine-arm . I've also rebased this PR on top of development, so it is ready for a re-review.

gilles-peskine-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -1743,7 +1745,7 @@
- * provided mbedtls/config.h contains the necessary inclusion.
+ * provided mbedtls/mbedtls_config.h contains the necessary inclusion.
  *
  * This feature is still experimental and is not ready for production since
```
Yes.

gilles-peskine-arm commented on docs/3.0-migration-guide.d/split_config.md:
```diff
@@ -0,0 +1,20 @@
+
+* In code, use `#include <mbedtls/build_info.h>`. Don't include `mbedtls/config.h` and don't refer to `MBEDTLS_CONFIG_FILE`.
+* In build tools, edit `mbedtls_config.h`, or edit `MBEDTLS_CONFIG_FILE` as before.
+* If you had a tool that parsed the library version from `include/mbedtls/version.h`, this has moved to `include/mbedtls/build_info.h`. From C code, both headers now define the `MBEDTLS_VERSION_xxx` macros.
```
No. It's markdown.

bensze01 commented:
@mstarzyk-mobica I've pushed some new commits that I think address all of your comments.

mstarzyk-mobica reviewed (approved):
Looks alright!
